### DISCUSSION TITLE
On customization points and `niebloids`

### DISCUSSION THREAD

**Discussion by nahueespinosa** [Ekumen-OS/beluga#44] [Ideas]
### Customization points and ADL

Customization points are hooks used by generic code that end-users can specialize to customize the behavior for their types. The STL provides plenty of examples: `begin`, `end`, `swap`, etc.

[Argument-dependent lookup](https://en.cppreference.com/w/cpp/language/adl) is a set of rules for looking up the unqualified function names in [function-call expressions](https://en.cppreference.com/w/cpp/language/operator_other), including implicit function calls to [overloaded operators](https://en.cppreference.com/w/cpp/language/operators) and is very relevant when implementing customization points.

If one were to make a call to a customization point function in generic code, the correct way to do it is the following:

```cpp
using std::swap;  // Make std::swap a candidate in name resolution
swap(a, b);       // Make an unqualified call and depend on ADL to find the right candidate (most specialized)
```

This is known as the _std-swap-two-step_ idiom and is used within the STL itself (see [`Swappable`](https://en.cppreference.com/w/cpp/named_req/Swappable)). The alternatives are incorrect (in a **generic context**) for the following reasons:

```cpp
std::swap(a, b);  // It's a qualified call, so ADL doesn't trigger and it won't find user-defined implementations
```
```cpp
swap(a, b);       // It won't find the standard implementation if a user-defined one doesn't exist
```

It's really a shame that the correct way to do this is not intuitive and, when a library author provides customization points based on pre-C++20 standard patterns, it's the user's responsibility to do the right thing.

### Niebloids!

Eric Niebler proposed [a solution](https://ericniebler.com/2014/10/21/customization-point-design-in-c11-and-beyond/) for this when designing the `range-v3` library (and in 2018, he also [suggested a name](https://twitter.com/ericniebler/status/1063850820569985024?s=20&t=6NnFjdP0lIPbUb-QJeTspQ) for it :sweat_smile:).

```cpp
namespace my_std {

namespace _detail {

struct _my_swap {
  template<typename T>
  void operator()(T& a, T& b) const {
    using std::swap;
    swap(a, b);
  }
};

} // namespace _detail

inline constexpr _detail::_my_swap my_swap{};  // since C++17's inline variables

} // namespace my_std
```

Now, instead of having a function, we have a global functor (called customization point object or CPO) that has the following properties:
- Cannot be found through ADL.
- Inhibits ADL (if this functor is found, no ADL is performed).
- The template parameters of the call operator cannot be explicitly specified.

The responsibility of doing the right thing is back in the hands of the library author, making the API safer. And as a bonus, they work really well with concept checking.

From C++20 onwards, we have CPOs in the standard library, i.e. see [`std::ranges::swap`](https://en.cppreference.com/w/cpp/utility/ranges/swap).

> Note the standard distinguishes CPOs (like ranges::begin) from niebloids. A niebloid is a non-customizable function that is never found by ADL. Both niebloids and CPOs are implemented as objects, but a niebloid has no user-defined customization.

### But wait, there is more...

- [P1895](https://github.com/cplusplus/papers/issues/643) tag_invoke: A general pattern for supporting customisable functions
- [P2547](https://github.com/cplusplus/papers/issues/1208) Language support for customisable functions

**Comment by nahueespinosa** [comment]
The current approach to static dispatch in `beluga` is to use type traits (i.e. `beluga::views::all`). This may be enough for our use case but CPOs are worth considering in the future if we want to provide more customization options.

**Comment by nahueespinosa** [comment]
P2547 has a very interesting [summary](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2547r1.html#the-problem-with-cpos-defined-in-ranges) of why ranges CPOs are not a panacea:

- This approach still generally relies on the same approach as raw ADL which has a single-global namespace for customisation-point names and so is susceptible to conflicts between libraries.
- Unable to generically customise and forward implementations in wrapper types.
- Lots of boiler-plate needed to define a CPO.
- The extra layer of indirection through the `operator()` overload inhibits copy-elision.

**Comment by hidmic** [comment]
> The extra layer of indirection through the operator() overload inhibits copy-elision.

Hmm, I don't follow why the compiler wouldn't be able to see right through.

**Comment by nahueespinosa** [comment]
> Hmm, I don't follow why the compiler wouldn't be able to see right through.

[Copy elision](https://en.cppreference.com/w/cpp/language/copy_elision) only works with pass-by-value semantics. If we're taking a reference, objects cannot be directly constructed into the storage where they would be copied/moved to. Even if we try to forward the reference by casting it to rvalue reference, that becomes an xvalue, not a prvalue.

Just to double check, I wrote this example ([godbolt link](https://godbolt.org/z/351YE5Mfa)):

Compiler flags: `-std=c++20 -O3`

```cpp
#include <string>
#include <iostream>

namespace _foo {
  void foo();
  struct _fn {
    template<typename T, typename V>
      requires requires (T& obj, V&& value) {
        foo(obj, (V&&)value);
      }
    void operator()(T& obj, V&& value) const {
      foo(obj, (V&&)value);
    }
  };
}

inline namespace _foo_cpo {
  inline constexpr _foo::_fn foo{};
}

struct string_wrapper {
  string_wrapper(std::string str_) : str{str_} {
      std::cout << "constructing\n";
  }
  string_wrapper(const string_wrapper& other) : str{other.str} {
      std::cout << "copying!\n";
  }
  string_wrapper(string_wrapper&& other) : str{std::move(other.str)} {
      std::cout << "moving!\n";
  }

  std::string str;
};

struct my_type {
  friend void foo(my_type& self, string_wrapper value) {
      std::cout << value.str;
  };
  friend void bar(my_type& self, string_wrapper value) {
      std::cout << value.str;
  };
};

int main() {
  my_type t;
  foo(t, string_wrapper(std::string{"hello\n"})); // will result in an extra call to
                                                  // string_wrapper's move constructor.
  bar(t, string_wrapper(std::string{"bye\n"}));
}
```

Output:
```
constructing
moving!
hello
constructing
bye
```

**Comment by hidmic** [comment]
Ah, I was thinking of RVO as applied to the return value of the CPO in an eligible scenario, not to temporaries used for argument passing. Yes, it's pass-by-value semantics or nothing.

**Comment by nahueespinosa** [comment]
Closed by #272
